Support private SuperPMI MCH file stores

BruceForstall opened the pull request:
In addition to the default Azure Storage SuperPMI location,
support private data stores via the new `-private_store` argument
to `replay` and `asmdiffs`. Private stores are typically file
system locations, or, on Windows, UNC paths.

There already exists the `-mch_files` argument, which allows you
to use any specified set of MCH files. However, when specifying
`-mch_files`, only those files are used and the normal stores are
ignored.

The private store mechanism is more convenient in some scenarios,
as it treats the specified private stores equivalently to the default
Azure Storage store.

In addition, both private stores and `-mch_files` handle downloading
ZIP files, un-ZIPping, and caching their contents.

Finally, a new top-level command `upload-private` is introduced to
help create private stores by specifying a set of MCH files to
use. It ZIP compresses the files to upload.

BruceForstall committed:
Support private SuperPMI MCH file stores

In addition to the default Azure Storage SuperPMI location,
support private data stores via the new `-private_store` argument
to `replay` and `asmdiffs`. Private stores are typically file
system locations, or, on Windows, UNC paths.

There already exists the `-mch_files` argument, which allows you
to use any specified set of MCH files. However, when specifying
`-mch_files`, only those files are used and the normal stores are
ignored.

The private store mechanism is more convenient in some scenarios,
as it treats the specified private stores equivalently to the default
Azure Storage store.

In addition, both private stores and `-mch_files` handle downloading
ZIP files, un-ZIPping, and caching their contents.

Finally, a new top-level command `upload-private` is introduced to
help create private stores by specifying a set of MCH files to
use. It ZIP compresses the files to upload.

BruceForstall commented:
@dotnet/jit-contrib PTAL

kunalspathak commented:
Will it be better if instead of introducing new top-level `upload-private`, we just add a command line option of `-private-store` to the existing `upload` command? That way, the meaning of `-private-store` (path to the private store) will be used uniformly in `replay`, `asmdiff`,  `download` and `upload`? For `upload`, you can add a check that either `-private-store` or `-az_storage_key` is specified but not both? That way we won't accidently upload anything private to azure storage?

kunalspathak commented:
Also, is it possible to add some kind of environment variable for `SPMI_LOCATION`. If I have multiple repos, then I can just set a system wide environment variable where the spmi collection, etc. should be downloaded and multiple repos can share those .mch files. Today, I have to remember passing `spmi_location SPMI_LOCATION` parameter.

BruceForstall commented:
> Will it be better if instead of introducing new top-level upload-private, we just add a command line option of -private-store to the existing upload command?

I made it different just to make it very clear how different it was, and to prevent "accidents".

> For upload, you can add a check that either -private-store or -az_storage_key is specified but not both? That way we won't accidently upload anything private to azure storage?

Technically, you could get az_storage_key from a persistent environment variable and still want to use upload-private, but maybe that should be flagged, anyway, if we're being paranoid.

> is it possible to add some kind of environment variable for SPMI_LOCATION

Possible (though unrelated), and a good idea. I have multiple enlistments and just end up downloading in multiple places, but I guess I haven't worried too much about download speed/time or cache space.

kunalspathak commented:
> I made it different just to make it very clear how different it was, and to prevent "accidents".

Can we then rename `-target_dir` of `upload-private` to be `-private-store` because that denotes the private storage location.

>  but maybe that should be flagged, anyway, if we're being paranoid.

Yes, lets add that validation.

kunalspathak commented:
> Possible (though unrelated), and a good idea. I have multiple enlistments and just end up downloading in multiple places, but I guess I haven't worried too much about download speed/time or cache space.

https://github.com/dotnet/runtime/pull/50298

BruceForstall commented:
@dotnet/jit-contrib Any other feedback on this?

kunalspathak commented:
> Technically, you could get az_storage_key from a persistent environment variable and still want to use upload-private, but maybe that should be flagged, anyway, if we're being paranoid.

Are we adding a flag to this?

BruceForstall commented:
> Are we adding a flag to this?

Do you mean, can I add an error if both az_storage_key is set and upload-private (or `upload --private_store`) is set? I could do that.

kunalspathak commented:
> Do you mean, can I add an error if both az_storage_key is set and upload-private (or `upload --private_store`) is set? I could do that.

Yes, that way we don't accidently upload things in wrong storage.

JulieLeeMSFT commented:
> add an error if both az_storage_key is set and upload-private (or upload --private_store) is set

Agreed to have this additional protection.

BruceForstall commented:
Updated. @dotnet/jit-contrib PTAL

BruceForstall committed:
Fix `-base_jit_path`

If `-base_jit_path` is specified on the command line, ensure it is a
full path before using it.
